### ISSUE TITLE
Request not being aborted after closing the connection is causing high CPU

### ISSUE THREAD

**Issue by arekpalinski** [dotnet/aspnetcore#52564]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

We experience an issue with unresponsiveness of RavenDB server (https://github.com/ravendb/ravendb) after a set of certain operations is cancelled. The operations in question are the streaming of query results (https://ravendb.net/docs/article-page/5.4/csharp/client-api/session/querying/how-to-stream-query-results).

The client application is creating 100 tasks and each of them is getting 2M records from a database. Under the covers each creates an HTTP request. The server handles the requests in a streaming manner and writes the results to `HttpContext.Response`.

The relevant server side code is here:
https://github.com/ravendb/ravendb/blob/917bbe60b099cd710de1f04cd27be8442251f790/src/Raven.Server/Documents/Handlers/Streaming/StreamingHandler.cs#L227-L267

Kestrel configuration:
https://github.com/ravendb/ravendb/blob/917bbe60b099cd710de1f04cd27be8442251f790/src/Raven.Server/RavenServer.cs#L185

When the streaming queries are handled have the following utilization:

- ~100 of connections ("18384" is pid of Raven.Server process):

```
C:\Users\arek>netstat -ano | find /c "18384"
112
```

- The CPU is 20% - 60% on my end (it's 12-core machine, 32GB of memory)

![image](https://github.com/dotnet/aspnetcore/assets/376551/54d49dd1-b2c6-42b0-92de-9b2618a391cb)

- It's ~9GB of managed memory

```
ravendb> stats
Showing stats, press any key to close...
    working set     | native mem      | managed mem     | mmap size          
 *  10.57 GBytes    | 29 MBytes       | 9.29 GBytes     | 4.54 GBytes
 ```

### The issue

The issue is happening once the client app is closed, Then we have very high CPU and memory usage is increasing. The server becomes unresponsive. After attaching the debugger we can see that the query results are still being written to the response (the underling stream is `Microsoft.AspNetCore.ResponseCompression.ResponseCompressionBody`) and all operations continue as usual. We have a cancellation token hooked up to `HttpContext.RequestAborted` but it's not cancelled:
https://github.com/ravendb/ravendb/blob/917bbe60b099cd710de1f04cd27be8442251f790/src/Raven.Server/Documents/DatabaseRequestHandler.cs#L140

- The number of connections drops immediately:

```
C:\Users\arek>netstat -ano | find /c "18384"
12
```

- CPU usage is over 90%

![image](https://github.com/dotnet/aspnetcore/assets/376551/246c905d-21f2-41de-b65d-0cb320b8c4f4)

- The process started to consume additional 2GB of managed memory

```
ravendb> stats
Showing stats, press any key to close...
    working set     | native mem      | managed mem     | mmap size       
 *  12.64 GBytes    | 28.29 MBytes    | 11.45 GBytes    | 4.54 GBytes
 ```

After some time (once all streaming operations completes I believe) the CPU get back to normal.

### Expected Behavior

- The cancellation token `HttpContext.RequestAborted` should be cancelled if the underlying connection no longer exists
- We expected to get some error on attempt to flush (we use default `KestrelServerOptions.Limits.MaxResponseBufferSize` which is 64KB) so we would stop the operation. Instead it appears to write to the memory so fast that it causes very high CPU usage.

### Steps To Reproduce

- checkout branch https://github.com/arekpalinski/ravendb/tree/streaming-issue
- start `Raven.Server` project
- create a database named `test`
- go to the database and import below .ravendump file (https://ravendb.net/docs/article-page/5.4/csharp/studio/database/tasks/import-data/import-data-file):
   - file to import: https://drive.google.com/file/d/10pbv43kI6q9KM9xRC91ybfy5LctURv_X/view?usp=sharing
   
   
- run project `Tryouts` which will create the actual requests
- stop it by pressing `q`

We have reproduce it locally on Win11. But the same was experienced on WinServer2022 and Ubuntu 22.04.

### .NET Version

7.0.14

**Comment by amcasey** [comment]
> We have a cancellation token hooked up to HttpContext.RequestAborted but it's not cancelled

Can you tell whether the token indicates that cancellation has occurred?  That is, whether the token is not updated or it's not consumed?

**Comment by amcasey** [comment]
Are you able to debug/profile to determine which code is executing between the disconnection and the release of resources?

Is this behavior new in 7.0?  (e.g. were you using 6.0 successfully before?)

Are you able to test on 8.0?

**Comment by arekpalinski** [comment]
Below is a sample stacktrace which shows that the query streaming operation is continuing its execution while the connection no longer exists:

```
>	Sparrow.dll!Sparrow.Json.AsyncBlittableJsonTextWriter.FlushAsync(System.Threading.CancellationToken token) Line 55	C#
 	Raven.Server.dll!Raven.Server.Documents.Queries.StreamJsonDocumentQueryResultWriter.AddResultAsync(Raven.Server.Documents.Document res, System.Threading.CancellationToken token) Line 57	C#
 	Raven.Server.dll!Raven.Server.Documents.Queries.StreamDocumentQueryResult.AddResultAsync(Raven.Server.Documents.Document result, System.Threading.CancellationToken token) Line 20	C#
 	Raven.Server.dll!Raven.Server.Documents.Indexes.Index.QueryInternal<Raven.Server.Documents.Queries.StreamDocumentQueryResult>(Raven.Server.Documents.Queries.StreamDocumentQueryResult resultToFill, Raven.Server.Documents.Queries.IndexQueryServerSide query, Raven.Server.Documents.Indexes.QueryOperationContext queryContext, bool pulseDocsReadingTransaction, Raven.Server.ServerWide.OperationCancelToken token) Line 3324	C#
 	[Resuming Async Method]	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 137	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.AsyncStateMachineBox<Raven.Server.Documents.Indexes.Index.<QueryInternal>d__243<Raven.Server.Documents.Queries.StreamDocumentQueryResult>>.MoveNext(System.Threading.Thread threadPoolThread) Line 133	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask) Line 135	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining) Line 162	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(object continuationObject) Line 2760	C#
 	Raven.Server.dll!Raven.Server.Documents.Queries.StreamDocumentQueryResult.AddResultAsync(Raven.Server.Documents.Document result, System.Threading.CancellationToken token) Line 25	C#
 	[Resuming Async Method]	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 137	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.AsyncStateMachineBox<Raven.Server.Documents.Queries.StreamDocumentQueryResult.<AddResultAsync>d__1>.MoveNext(System.Threading.Thread threadPoolThread) Line 133	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask) Line 135	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining) Line 162	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(object continuationObject) Line 2760	C#
 	Raven.Server.dll!Raven.Server.Documents.Queries.StreamJsonDocumentQueryResultWriter.AddResultAsync(Raven.Server.Documents.Document res, System.Threading.CancellationToken token) Line 58	C#
 	[Resuming Async Method]	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 137	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.AsyncStateMachineBox<Raven.Server.Documents.Queries.StreamJsonDocumentQueryResultWriter.<AddResultAsync>d__8>.MoveNext(System.Threading.Thread threadPoolThread) Line 133	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask) Line 135	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining) Line 162	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(object continuationObject) Line 2760	C#
 	Sparrow.dll!Sparrow.Json.AsyncBlittableJsonTextWriter.FlushAsync(System.Threading.CancellationToken token) Line 58	C#
 	[Resuming Async Method]	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 137	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<int>.AsyncStateMachineBox<Sparrow.Json.AsyncBlittableJsonTextWriter.<FlushAsync>d__4>.MoveNext(System.Threading.Thread threadPoolThread) Line 133	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask) Line 135	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining) Line 162	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(object continuationObject) Line 2760	C#
 	Microsoft.AspNetCore.ResponseCompression.dll!Microsoft.AspNetCore.ResponseCompression.ResponseCompressionBody.WriteAsync(byte[] buffer, int offset, int count, System.Threading.CancellationToken cancellationToken)	Unknown
 	[Resuming Async Method]	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 137	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.AsyncStateMachineBox<Microsoft.AspNetCore.ResponseCompression.ResponseCompressionBody.<WriteAsync>d__40>.MoveNext(System.Threading.Thread threadPoolThread) Line 133	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask) Line 135	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining) Line 162	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(object continuationObject) Line 2760	C#
 	Microsoft.AspNetCore.ResponseCompression.dll!Microsoft.AspNetCore.ResponseCompression.ResponseCompressionBody.WriteAsync(System.ReadOnlyMemory<byte> buffer, System.Threading.CancellationToken cancellationToken)	Unknown
 	[Resuming Async Method]	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 137	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.AsyncStateMachineBox<Microsoft.AspNetCore.ResponseCompression.ResponseCompressionBody.<WriteAsync>d__41>.MoveNext(System.Threading.Thread threadPoolThread) Line 133	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask) Line 135	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining) Line 162	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(object continuationObject) Line 2760	C#
 	System.IO.Compression.dll!System.IO.Compression.DeflateStream.WriteAsyncMemory.__Core|65_0(System.ReadOnlyMemory<byte> buffer, System.Threading.CancellationToken cancellationToken)	Unknown
 	[Resuming Async Method]	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 137	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.AsyncStateMachineBox<System.IO.Compression.DeflateStream.<<WriteAsyncMemory>g__Core|65_0>d>.MoveNext(System.Threading.Thread threadPoolThread) Line 133	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask) Line 135	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining) Line 162	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(object continuationObject) Line 2760	C#
 	System.IO.Compression.dll!System.IO.Compression.DeflateStream.WriteDeflaterOutputAsync(System.Threading.CancellationToken cancellationToken)	Unknown
 	[Resuming Async Method]	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 137	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.Threading.Tasks.VoidTaskResult>.AsyncStateMachineBox<System.IO.Compression.DeflateStream.<WriteDeflaterOutputAsync>d__66>.MoveNext(System.Threading.Thread threadPoolThread) Line 133	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask) Line 135	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.RunOrScheduleAction(System.Action action, bool allowInlining) Line 162	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.Task.RunContinuations(object continuationObject) Line 2760	C#
 	Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.PipeWriterHelpers.TimingPipeFlusher.TimeFlushAsyncAwaited(System.Threading.Tasks.ValueTask<System.IO.Pipelines.FlushResult> pipeFlushTask, Microsoft.AspNetCore.Server.Kestrel.Core.MinDataRate minRate, Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.IHttpOutputAborter outputAborter, System.Threading.CancellationToken cancellationToken)	Unknown
 	[Resuming Async Method]	
 	System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 137	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.AsyncTaskMethodBuilder<System.IO.Pipelines.FlushResult>.AsyncStateMachineBox<Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.PipeWriterHelpers.TimingPipeFlusher.<TimeFlushAsyncAwaited>d__9>.MoveNext(System.Threading.Thread threadPoolThread) Line 133	C#
 	System.Private.CoreLib.dll!System.Runtime.CompilerServices.TaskAwaiter.OutputWaitEtwEvents.AnonymousMethod__12_0(System.Action innerContinuation, System.Threading.Tasks.Task innerTask) Line 135	C#
 	System.Private.CoreLib.dll!System.Threading.Tasks.AwaitTaskContinuation.System.Threading.IThreadPoolWorkItem.Execute() Line 94	C#
 	System.Private.CoreLib.dll!System.Threading.ThreadPoolWorkQueue.Dispatch() Line 692	C#
 	System.Private.CoreLib.dll!System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() Line 1177	C#
 	[Async Call Stack]	
 	[Async] Raven.Server.dll!Raven.Server.Documents.Indexes.Index.StreamQuery(Microsoft.AspNetCore.Http.HttpResponse response, Raven.Server.Documents.Queries.IStreamQueryResultWriter<Raven.Server.Documents.Document> writer, Raven.Server.Documents.Queries.IndexQueryServerSide query, Raven.Server.Documents.Indexes.QueryOperationContext queryContext, Raven.Server.ServerWide.OperationCancelToken token) Line 3099	C#
 	[Async] Raven.Server.dll!Raven.Server.Documents.Queries.StaticIndexQueryRunner.ExecuteStreamQuery(Raven.Server.Documents.Queries.IndexQueryServerSide query, Raven.Server.Documents.Indexes.QueryOperationContext queryContext, Microsoft.AspNetCore.Http.HttpResponse response, Raven.Server.Documents.Queries.IStreamQueryResultWriter<Raven.Server.Documents.Document> writer, Raven.Server.ServerWide.OperationCancelToken token) Line 51	C#
 	[Async] Raven.Server.dll!Raven.Server.Documents.Queries.QueryRunner.ExecuteStreamQuery(Raven.Server.Documents.Queries.IndexQueryServerSide query, Raven.Server.Documents.Indexes.QueryOperationContext queryContext, Microsoft.AspNetCore.Http.HttpResponse response, Raven.Server.Documents.Queries.IStreamQueryResultWriter<Raven.Server.Documents.Document> writer, Raven.Server.ServerWide.OperationCancelToken token) Line 119	C#
 	[Async] Raven.Server.dll!Raven.Server.Documents.Handlers.Streaming.StreamingHandler.StreamQueryPost() Line 287	C#
 	[Async] Raven.Server.dll!Raven.Server.Routing.RequestRouter.HandlePath(Raven.Server.Web.RequestHandlerContext reqCtx) Line 378	C#
 	[Async] Raven.Server.dll!Raven.Server.RavenServerStartup.RequestHandler(Microsoft.AspNetCore.Http.HttpContext context) Line 178	C#
 	[Async] Microsoft.AspNetCore.ResponseCompression.dll!Microsoft.AspNetCore.ResponseCompression.ResponseCompressionMiddleware.InvokeCore(Microsoft.AspNetCore.Http.HttpContext context)	Unknown
 	[Async] Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests<Microsoft.AspNetCore.Hosting.HostingApplication.Context>(Microsoft.AspNetCore.Hosting.Server.IHttpApplication<Microsoft.AspNetCore.Hosting.HostingApplication.Context> application)	Unknown
 	[Async] Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequestsAsync<Microsoft.AspNetCore.Hosting.HostingApplication.Context>(Microsoft.AspNetCore.Hosting.Server.IHttpApplication<Microsoft.AspNetCore.Hosting.HostingApplication.Context> application)	Unknown
 	[Async] Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.HttpConnection.ProcessRequestsAsync<Microsoft.AspNetCore.Hosting.HostingApplication.Context>(Microsoft.AspNetCore.Hosting.Server.IHttpApplication<Microsoft.AspNetCore.Hosting.HostingApplication.Context> httpApplication)	Unknown
 	[Async] Microsoft.AspNetCore.Server.Kestrel.Core.dll!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.KestrelConnection<Microsoft.AspNetCore.Connections.ConnectionContext>.ExecuteAsync()	Unknown

```

Then we can chek in the debugger that the underlaying HTTP context's `RequestAborted` token doesn't indicate the cancellation:

![image](https://github.com/dotnet/aspnetcore/assets/376551/202762c2-aa8a-4957-a908-82382e555703)

This is the first time we noticed an issue like that. I don't know if this is new dotnet 7.0 behavior.

**Comment by arekpalinski** [comment]
The same is experienced on .net 8.0:

![image](https://github.com/dotnet/aspnetcore/assets/376551/9a2be5c5-b575-4ce3-ad96-c2887cc5f384)


We can even see with the usage of `netstat` that there is no connection for the port `57886`:

![image](https://github.com/dotnet/aspnetcore/assets/376551/c1b5e878-56b2-4c20-819e-94bf41627690)

**Comment by amcasey** [comment]
Thanks for the updates!  I'll see if I can figure out exactly what triggers `RequestAborted`.

**Comment by amcasey** [comment]
[This](https://github.com/dotnet/aspnetcore/blob/ffa0a028464e13d46aaec0c5ad8de0725a4d5aa5/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L272) appears to be the relevant implementation of `RequestAborted` and [this](https://github.com/dotnet/aspnetcore/blob/ffa0a028464e13d46aaec0c5ad8de0725a4d5aa5/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L486) appears to be the method that triggers it (unless it's explicitly [clobbered](https://github.com/dotnet/aspnetcore/blob/ffa0a028464e13d46aaec0c5ad8de0725a4d5aa5/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L304), which I assume it isn't in your case).

It might be interesting to put a breakpoint in `CancelRequestAbortedToken` to confirm that it's not being called.  (I do appreciate that you provided repro steps, but setting up a DB server populated from an opaque binary blob is not something we'd generally do for an investigation.)

**Comment by amcasey** [comment]
Since you seem to have a convenient repro, I'd also be tempted to turn the logging for `Microsoft.AspNetCore` up to `Debug` to get some messages about connection lifetimes.

**Comment by arekpalinski** [comment]
Next, I repeated the same test with `Debug` log level. Please find attached the logs:
[2023-12-08.000.log](https://github.com/dotnet/aspnetcore/files/13614240/2023-12-08.000.log)

**Comment by arekpalinski** [comment]
> I do appreciate that you provided repro steps, but setting up a DB server populated from an opaque binary blob is not something we'd generally do for an investigation

The binary blob isn't necessary. I have attached it as it's easier to just import it to RavenDB. Although it's just a sample built-in data set that can be easily extended to be ~2M of documents.

You can achieve the same data set by doing:

1. Download RavenDB Server from https://ravendb.net/download (vesion 5.4 / 6.0 doesn't matter)
2. Open RavenDB Studio (http://localhost:8080) and create "test" database
3. Go to "test" database and create "Sample data":
![image](https://github.com/dotnet/aspnetcore/assets/376551/6d9ff3b5-7ea8-47ab-8b58-09067e1ae7e4)
4.  Go to "Documents" -> "Patch" and execute the following script which will create more "Orders" documents:

```
from "Orders" update {
    for (var i = 0; i < 20; i++)
    {
        put("orders/", this)
    }
}
```
![image](https://github.com/dotnet/aspnetcore/assets/376551/84719195-f3cc-4ca8-87a3-657a2c0e163e)

Execute it a few times until you have ~2M of Orders documents

![image](https://github.com/dotnet/aspnetcore/assets/376551/89070e82-834c-45cb-b6ef-1678e2b6aa5f)

**Comment by amcasey** [comment]
Thanks for the logs!

Here's the isolated trace of a single connection
```
2023-12-08T12:55:14.1208847Z, 11, Operations, Microsoft.AspNetCore.Server.Kestrel, Connections, Debug, Connection id "0HMVNTJK66HEJ" accepted.
2023-12-08T12:55:14.1209107Z, 9, Operations, Microsoft.AspNetCore.Server.Kestrel, Connections, Debug, Connection id "0HMVNTJK66HEJ" started.
2023-12-08T12:55:16.4657524Z, 22, Operations, Microsoft.AspNetCore.Server, Kestrel, Debug, Connection id "0HMVNTJK66HEJ", Request id "0HMVNTJK66HEJ:00000001": started reading request body.
2023-12-08T12:55:29.7267191Z, 83, Operations, Microsoft.AspNetCore.Server.Kestrel.Transport, Sockets, Debug, Connection id "0HMVNTJK66HEJ" reset.
2023-12-08T12:55:29.7267267Z, 83, Operations, Microsoft.AspNetCore.Server.Kestrel.Transport, Sockets, Debug, Connection id "0HMVNTJK66HEJ" sending RST because: "An existing connection was forcibly closed by the remote host."
2023-12-08T12:55:29.7430104Z, 87, Operations, Microsoft.AspNetCore.Server.Kestrel, Connections, Debug, Connection id "0HMVNTJK66HEJ" disconnecting.
2023-12-08T12:55:57.2625475Z, 14, Operations, Microsoft.AspNetCore.Server, Kestrel, Debug, Connection id "0HMVNTJK66HEJ", Request id "0HMVNTJK66HEJ:00000001": done reading request body.
2023-12-08T12:55:57.2626354Z, 14, Operations, Microsoft.AspNetCore.Server.Kestrel, Connections, Debug, Connection id "0HMVNTJK66HEJ" stopped.
```

The connection is reset after 13 seconds, but the request body continues to be read for another 28 seconds.  That's suspicious, but may just be a consequence of the CancellationToken not firing (i.e. it may be downstream from the real problem).

**Comment by amcasey** [comment]
> The binary blob isn't necessary. I have attached it as it's easier to just import it to RavenDB. Although it's just a sample built-in data set that can be easily extended to be ~2M of documents.

Clearly the entire codebase can't be taken out of the equation, but it seems like it shouldn't matter what work is being done while waiting for cancellation?  That is, there's no particular reason to read from the database at all, as long as request processing doesn't terminate before the client closes the connection?

**Comment by amcasey** [comment]
The repro seems like it could be simplified.  I can't tell from your description whether the client or the server is doing the database fetches.  If it's the client, sending dummy data (e.g. all zeroes) should suffice.  If it's the server, a simple `Task.Delay` or `Thread.Sleep` should suffice.  I don't believe there's any need to have multiple concurrent connections.  (This could be wrong, but let's start with one and see what happens.)

My initial guess is that `SocketConnection.DoReceive` does too much work between checks of the cancellation status and, therefore, calls `FireConnectionClosed` too late.

**Comment by maciejaszyk** [comment]
It doesn't reproduce the same log commands as the example above, but if you have high CPU usage, the tasks are still running even if Netstats shows that sockets are closed. If you comment out the `CpuUsage` from the `StreamingController` and use `Thread.Sleep`, tasks should end pretty soon.

Repro can be found here: https://github.com/maciejaszyk/ravendb/tree/KestrelReproduction/Reproduction

**Comment by amcasey** [comment]
Unfortunately, I've had less time to look at this than I was hoping due to an ongoing series of urgent requests.  Since it's so late in the year, I may not have time to pick this back up until January.  Thanks for your patience - I wish I had better news.

**Comment by arekpalinski** [comment]
Thanks for letting us know. Looking forward to hear back from you in 2024.

**Comment by arekpalinski** [comment]
@amcasey did you have a chance to take a look at this?

**Comment by amcasey** [comment]
Nope, tomorrow is my first day back. 😉

**Comment by amcasey** [comment]
> It doesn't reproduce the same log commands as the example above, but if you have high CPU usage, the tasks are still running even if Netstats shows that sockets are closed. If you comment out the `CpuUsage` from the `StreamingController` and use `Thread.Sleep`, tasks should end pretty soon.
> 
> Repro can be found here: https://github.com/maciejaszyk/ravendb/tree/KestrelReproduction/Reproduction

I'm tacitly assuming the client was intended to pass `HttpCompletionOption.ResponseHeadersRead` so that it doesn't wait indefinitely for the response to be available.

**Comment by amcasey** [comment]
Things get a lot better when I thread the cancellation token into `WriteAsync`, `CpuUsage`, and `FlushAsync` in the controller, but it's still doing work for a while after the client disconnects.

**Comment by amcasey** [comment]
AFAICT, `FireConnectionClosed` is called in a timely manner, but `CancelConnectionClosedToken` is not.  Context: https://github.com/dotnet/aspnetcore/blob/70abdab0a5969fd7746479c1fae4697bce2dbc92/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs#L357

**Comment by amcasey** [comment]
I don't have a smoking gun, but my assessment at this point is that, with 100 threads each grabbing the CPU uninterruptably for 3 seconds at a time, it might take a while for a thread pool thread to get a turn to actually fire the cancellation.  This is exacerbated by the fact that the connection-close/shutdown/cancellation flow yields in several places.  In addition to `FireConnectionClosed` (above), there's `CancelRequestAbortedToken`: 
https://github.com/dotnet/aspnetcore/blob/70abdab0a5969fd7746479c1fae4697bce2dbc92/src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs#L504

**Comment by amcasey** [comment]
Here's the sequence of events, as I currently understand it, for a single HTTP/1.1 connection:

The client gracefully closes the connection by sending a TCP `FIN`
In `SocketConnection.DoReceive`, `SocketReceiver.ReceiveAsync` returns a non-error result with 0 `BytesTransferred`
`DoReceive` calls `SocketConnection.FireConnectionClosed`
`FireConnectionClosed` _schedules_ `SocketConnection.CancelConnectionClosedToken`
`CancelConnectionClosedToken` cancels `_connectionClosedTokenSource`
`HttpConnection.OnConnectionClosed` is registered as a callback of such a `CancellationToken`
`OnConnectionClosed` calls `Http1Connection.OnInputOrOutputCompleted` (the `IRequestProcessor` version)
`OnInputOrOutputCompleted` calls `HttpProtocol.CancelRequestAbortedToken`
`CancelRequestAbortedToken` _schedules_ `HttpProtocol.CancelRequestAbortedTokenCallback`
`CancelRequestAbortedTokenCallback` cancels `_abortedCts` (modulo some special cases that don't seem to apply)
`StreamingController` is checking for cancellation of this token via `HttpContext.RequestAborted`

**Comment by amcasey** [comment]
TL;DR: connection cancellation, surfaced as `HttpContext.RequestAborted`, depends on being able to schedule notification actions to the thread pool.  If the thread pool is busy with jobs that never yield (e.g. with `await Task.Yield()`), then the notifications will not be delivered.

@arekpalinski Can you increase the minimum number of worker threads and/or perform more yields in your controller?

**Comment by arekpalinski** [comment]
@amcasey We will check that next week. Thank you

**Comment by maciejaszyk** [comment]
@amcasey thanks a lot for your investigation. I ran multiple tests on our repro (RavenDB) and seems that increasing min worker threads did the trick. Do you have any recommendations for what the minimum values should be for applications that have a lot of requests/works? (RavenDB is a database, and we expect to have multiple connections/threads)
